targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/serial_api.c

```diff
@@ -0,0 +1,378 @@
+void uart3_irq() {uart_irq(CMSDK_UART2->INTSTATUS & 0x3, 3, (CMSDK_UART_TypeDef*)CMSDK_UART2);}
+
+void serial_irq_handler(serial_t *obj, uart_irq_handler handler, uint32_t id) {
+    irq_handler = handler;
+    uart_data[obj->index].serial_irq_id = id;
+}
+
+static void serial_irq_set_internal(serial_t *obj, SerialIrq irq, uint32_t enable) {
+    IRQn_Type irq_n = (IRQn_Type)0;
+    uint32_t vector = 0;
+    switch ((int)obj->uart)
+    {
+        case UART_0:
+            irq_n = UART0_IRQn;
+            vector = (uint32_t)&uart0_irq;
+        break;
+        case UART_1:
+            irq_n = UART1_IRQn;
+            vector = (uint32_t)&uart1_irq;
+        break;
+        case UART_2:
+        {
+            irq_n = UART2_IRQn;
+            vector  = (uint32_t)&uart2_irq;
+        }
+        break;
+        case UART_3:
+        {
+            irq_n = UART3_IRQn;
+            vector  = (uint32_t)&uart3_irq;
+        }
+        break;
+    }
+
+    if (enable)
+    {
+        if (irq == TxIrq)
+        {
+            /* set TX interrupt enable in CTRL REG */
+            obj->uart->CTRL |= CMSDK_UART_CTRL_TXIRQEN_Msk;
+        }
+        else
+        {
+            /* set Rx interrupt on in CTRL REG */
+            obj->uart->CTRL |= CMSDK_UART_CTRL_RXIRQEN_Msk;
+        }
+        NVIC_SetVector(irq_n, vector);
+        NVIC_EnableIRQ(irq_n);
+    }
+    else if ((irq == TxIrq) || (uart_data[obj->index].rx_irq_set_api + uart_data[obj->index].rx_irq_set_flow == 0))
+    {   /*      Disable   IRQ   */
+        int all_disabled = 0;
+        SerialIrq other_irq = (irq == RxIrq) ? (TxIrq) : (RxIrq);
+
+        obj->uart->CTRL &= ~(1 << (irq + 2));
+
+        all_disabled = (obj->uart->CTRL & (1 << (other_irq + 2))) == 0;
+
+        if (all_disabled)
+        {
+            NVIC_DisableIRQ(irq_n);
+        }
+    }
+}
+
+void serial_irq_set(serial_t *obj, SerialIrq irq, uint32_t enable) {
+    if (RxIrq == irq)
+        uart_data[obj->index].rx_irq_set_api = enable;
+    serial_irq_set_internal(obj, irq, enable);
+}
+
+/******************************************************************************
+ * READ/WRITE
+ ******************************************************************************/
+int serial_getc(serial_t *obj) {
+    while (serial_readable(obj) == 0);
+    int data = obj->uart->DATA;
+    return data;
+}
+
+void serial_putc(serial_t *obj, int c) {
+    while (serial_writable(obj));
+    obj->uart->DATA = c;
+}
+
+int serial_readable(serial_t *obj) {
+    return obj->uart->STATE & 0x2;
+}
+
+int serial_writable(serial_t *obj) {
+    return obj->uart->STATE & 0x1;
+}
+
+void serial_clear(serial_t *obj) {
+    obj->uart->DATA = 0x00;
+}
+
+void serial_pinout_tx(PinName tx) {
+    pinmap_pinout(tx, PinMap_UART_TX);
+}
+
+void serial_break_set(serial_t *obj) {
+}
+
+void serial_break_clear(serial_t *obj) {
+}
+void serial_set_flow_control(serial_t *obj, FlowControl type, PinName rxflow, PinName txflow) {
+}
```

Comment:
The CMSDK APB UART doesn't support serial break and flow control. Ref. DDI0479C_cortex_m_system_design_kit_r1p0_trm.pdf. However, one comment and mbed error will be added in this function.